<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <head>
    <title>Release Notes of SBMLsqueezer version 1.3</title>
  </head>
  <body>
    <h1>Release Notes of SBMLsqueezer version 1.3</h1>
    <h2>Download of SBMLsquezeezer</h2>
    You can always download the latest version of SBMLsqueezer at
    <a href="http://www.ra.cs.uni-tuebingen.de/software/SBMLsqueezer/downloads">
    http://www.ra.cs.uni-tuebingen.de/software/SBMLsqueezer/downloads</a>.
    
    <h2>New Features</h2>
    
    <h3>A New and Purely Java-Based SBML Data Structure</h3>
    
    <p>SBMLsqueezer's data structure was completely changed: Earlier versions
      were based on String concatenation to create kinetic formulas with all
      participating species and parameters. Now SBMLsqueezer is based on an
      abstract syntax tree representation of the formula and also creates 
      <span style="font-family:Courier">Parameter</span> objects rather than simply
      their identifier.</p>
    <p>This new data structure is actually an almost complete Java implementation
      of SBML and has therefore become a separate project, JSBML, available at
      <a href="http://sourceforge.net/projects/jsbml">
      http://sourceforge.net/projects/jsbml</a>.
      On the SBML homepage, you can find a separate mailing list about this
      project:
      <a href="http://sbml.org/Forums">http://sbml.org/Forums</a></p>
    <p>With JSBML at hand, SBMLsqueezer does no longer manipulate CellDesigner's
      plug-in Objects directly. If you start SBMLsqueezer from CellDesigner's
      plug-in menu or its reaction context menu, SBMLsqueezer will copy all
      required parts of the model into JSBML objects. Only this copy of the model
      is manipulated. By clicking on OK or Apply, all changes will be synchronized
      with CellDesigner's original data structures.</p>
    
    <h3>Stand-Alone Mode</h3>
    
    <p>SBMLsqueezer now also runs in a stand-alone mode. With the
      help of libSBML, valid SBML files can be read and copied into the JSBML
      data objects. JSBML now mirrors the complete definition of SBML up to Level 2
      Version 4. As in the CellDesigner plug-in mode, SBMLsqueezer only manipulates
      JSBML objects. By clicking on save, changes are written into the SBML
      file.</p>
    <p>Command-line mode: All features of SBMLsqueezer are now also available
      from the command line. This allows users to write shell or batch scripts that
      access SBMLsqueezer's functions.</p>
    <p>The Application Programming Interface (API) of SBMLsqueezer has been simplified.
      There is now just one function, <span style="font-family:Courier">squeeze</span>
      in the main class <span style="font-family:Courier">SBMLsqueezer</span> that
      can be called to read in a model from a valid SBML file and to write the result,
      i.e., an SBML model including new 
      <span style="font-family:Courier">KineticLaws</span>, new 
      <span style="font-family:Courier">Parameters</span>, new or adapted
      <span style="font-family:Courier">UnitDefinitions</span>, into the given out
      file. Before creating kinetic formulas for the model, the user probably wants
      to adjust SBMLsqueezer's configuration. To this end, the methods <span
      style="font-family:Courier">set(CfgKey, &#60;boolean|double|int|String&#62;)</span>
      in class <span style="font-family:Courier">SBMLsqueezer</span> allow the
      programmer to set any configuration key and value pair, similar to the command-line
      mode. Here we give a minimal example of how to use this new API:
      <table border="0" align="center" cellspacing="10" cellpadding="5"><tr><td><code>
      public static void main(String[] args) {<br>
      &#160;&#160;// Initialize SBMLsqueezer with appropriate SBML readers/writers<br>
      &#160;&#160;SBMLsqueezer squeezer = new SBMLsqueezer(new LibSBMLReader(), new LibSBMLWriter());<br>
      &#160;&#160;// Configure SBMLsqueezer<br>
      &#160;&#160;squeezer.set(CfgKeys.OPT_ALL_REACTIONS_ARE_ENZYME_CATALYZED, true);<br>
      &#160;&#160;squeezer.set(CfgKeys.OPT_DEFAULT_COMPARTMENT_INITIAL_SIZE, 1.0);<br>
      &#160;&#160;squeezer.set(CfgKeys.POSSIBLE_ENZYME_RNA, true);<br>
      &#160;&#160;squeezer.set(CfgKeys.KINETICS_UNI_UNI_TYPE, MichaelisMenten.class.getName());<br>
      &#160;&#160;squeezer.set(CfgKeys.KINETICS_OTHER_ENZYME_REACTIONS, ConvenienceKinetics.class.getName());<br>
      &#160;&#160;try {<br>
      &#160;&#160;&#160;&#160;// Create kinetic equations, parameters, units etc. and save the result<br>
      &#160;&#160;&#160;&#160;// args contains infile and outfile path<br>
      &#160;&#160;&#160;&#160;squeezer.squeeze(args[0], args[1]);<br>
      &#160;&#160;} catch (Throwable e) {<br>
      &#160;&#160;&#160;&#160;e.printStackTrace();<br>
      &#160;&#160;}<br>
      }<br>
      </code></td></tr></table>
    </p>
    <p>Since SBMLsqueezer still requires an installation of libSBML on the user's
      system, which can sometimes become problematic, a more convenient way of using
      SBMLsqueezer has now been made available: the SBMLsqueezer web application does
      not require any local installation and is freely available at <a 
      href="http://webservices.cs.uni-tuebingen.de">http://webservices.cs.uni-tuebingen.de</a>.
      In this framework, workflows can be created in which SBMLsqueezer can directly
      be linked to the full version of <a 
      href="http://www.ra.cs.uni-tuebingen.de/software/SBML2LaTeX">SBML2LaTeX</a>.
      The full version of SBML2LaTeX contains many additional features that are
      not included in SBMLsqueezer's version of SBML2LaTeX. Hence, it is highly
      recommended to make use of the online version of SBML2LaTeX.</p>
    
    <h3>Configuration</h3>
    
    <p>SBMLsqueezer now memorizes every setting and changed parameters
      from the Graphical User Interface (GUI) whether in CellDesigner plug-in or 
      stand-alone mode. In the stand-alone case, SBMLsqueezer even memorizes given
      command-line arguments. These settings are read from a configuration file at
      the next start. This is especially convenient for open and save directories.
      At any time, the user can change all settings to the defaults by clicking 
      the designated button in the preferences dialog.</p>
    
    <h3>Improved Systems Biology Ontology Support</h3>  
    
    <p>SBMLsqueezer is now fully based on the <a href="http://www.ebi.ac.uk/sbo/">
      Systems Biology Ontology (SBO)</a>. The JSBML copy of CellDesigner's plug-in
      data structures maps all CellDesigner-specific annotations to corresponding
      SBO terms. Afterwards these are mapped back. SBO term &#8220;enzymatic 
      catalyst&#8221; (<span style="font-family:Courier">SBO:0000460</span>) was 
      created because it is required by SBMLsqueezer to distinguish enzymes and other
      catalysts, such as certain anorganic ions. SBMLsqueezer's stand-alone version
      interpretes SBO terms that can be found in the model and changes the annotation
      of <span style="font-family:Courier">ModifierSpeciesReference</span>s from
      &#8220;catalyst&#8221; (<span style="font-family:Courier">SBO:0000013</span>)
      to &#8220;enzymatic catalyst&#8221; if the SBO term of the corresponding <span 
      style="font-family:Courier">Species</span> is one of those belonging to the
      family of materials considered as enzymes (SBMLsqueezer still offers check
      boxes to indicate certain kinds of species as enzymes).</p>
    <p>Earlier versions of SBMLsqueezer indicated the SBO term identifiers of newly
      created kinetic equations if these were already defined in SBO. However, this
      was more or less just for information to the user and could neither be saved
      nor be considered during any other processes. This version of SBMLsqueezer is
      able to not only take SBO terms into account while creating rate laws, but can
      even save SBO terms in the resulting SBML file. However, in CellDesigner plug-in
      mode, SBMLsqueezer is not yet able to save SBO term ids in CellDesigner data
      objects due to the missing functionality in its API.</p>
    <p>SBMLsqueezer now contains an own parser to identify the correct SBO term ids
      and attributes that are required to annotate all newly created objects.</p>
    <p>Whenever possible, SBMLsqueezer annotates newly created objects, such as
      <span style="font-family:Courier">Parameter</span>s or <span 
      style="font-family:Courier">UnitDefinition</span>s, with SBO identifiers.</p>
      
    <h3>MIRIAM Support</h3>
    
    <p>In contrast to earlier versions, SBMLsqueezer now also interprets and uses
    MIRIAM tags. The new command-line and configuration key option <span 
    style="font-family:Courier">OPT_IGNORE_THESE_SPECIES_WHEN_CREATING_LAWS</span>
    takes a list of comma-separated MIRIAM identifiers as value. Species annotated
    with BQB_IS and the given MIRIAM resource are then ignored when creating rate
    equations. This is, for instance, useful if small molecules such as water take
    part in a reaction. The concentration of water can hardly be measured and it is
    assumed to occur abundantly within the cell. By default, the list of species whose
    influence is to be neglected when creating new kinetic equations contains
    the KEGG compound identifiers for the following substances:
    <ul>
    <li><span style="font-family:Courier">C00001</span>: water</li>
    <li><span style="font-family:Courier">C00038</span>: Zn<sup>2+</sup></li>
    <li><span style="font-family:Courier">C00070</span>: Cu<sup>2+</sup></li>
    <li><span style="font-family:Courier">C00076</span>: Ca<sup>2+</sup></li>
    <li><span style="font-family:Courier">C00080</span>: proton</li>
    <li><span style="font-family:Courier">C00175</span>: Co<sup>2+</sup></li>
    <li><span style="font-family:Courier">C00238</span>: K<sup>+</sup></li>
    <li><span style="font-family:Courier">C00282</span>: H<sub>2</sub></li>
    <li><span style="font-family:Courier">C00291</span>: Ni<sup>2+</sup></li>
    <li><span style="font-family:Courier">C01327</span>: Cl<sup>-</sup></li>
    <li><span style="font-family:Courier">C01528</span>: H<sub>2</sub>Se</li>
    <li><span style="font-family:Courier">C14818</span>: Fe(II)</li>
    <li><span style="font-family:Courier">C14819</span>: Fe(III)</li>
    </ul></p>
    
    <h3>New Kinetic Equations and Extensibility</h3>

    <p>In earlier versions the inclusion of new kinetic equations into SBMLsqueezer
      was not an easy task. After writing a class that actually generates the new
      rate law, a rule had to be implemented of when to make this rate law available.
      Furthermore, the GUI had to be changed to include this new formula. Summarized,
      several changes in the program code were necessary to include additional rate
      laws.</p>
    <p>Now SBMLsqueezer is based on Java reflection, i.e., initially SBMLsqueezer does
      not know any one of its available kinetic equations. These are all loaded when
      initializing SBMLsqueezer. Currently, twelve interfaces define the properties 
      of all kinetic equations. The rules of when select which kind of rate equation
      only make use of the implemented interfaces. Hence, new kinetic equations can
      be incorporated into SBMLsqueezer simply by implementing designated interfaces.
      Since SBMLsqueezer is an open-source project, users can customize it easily and
      extend it with desired rate laws.</p>
    <p>Much effort was put into the extension of SBMLsqueezer with additional rate
      laws for gene-regulatory processes. Earlier versions of SBMLsqueezer only provided
      the Hill equation variant suggested by Hinze et al. 2007. This equation has now
      been renamed to Hill-Hinze equation to highlight that it is actually a modification
      of the original rate law. Only in special cases this equation equals the traditional
      Hill equation (SBMLsqueezer will highlight these cases). Now nine additional
      rate equations were included:<ul>
      <li>Additive Model Linear: A generalized super class of all other additive rate
        laws for gene-regulatory processes.</li>
      <li>Additive Model Non-Linear: defined in the paper &#8220;The NetGenerator 
        Algorithm: Reconstruction of Gene Regulatory Networks&#8221; of T&#246;pfer, 
        S.; Guthke, R.; Driesch, D.; W&#246;tzel, D., and Pfaff 2007</li>
      <li>H-Systems: Spieth, C.; Hassis, N.; Streichert, F.; Supper, J.; Beyreuther, K.,
        and Zell, A. &#8220;Comparing Mathematical Models on the Problem of Network
        Inference&#8221;, 2006.</li>
      <li>NetGenerator Linear Model: &#8220;The NetGenerator Algorithm: Reconstruction
        of Gene Regulatory Networks&#8221; of T&#246;pfer, S.; Guthke, R.; Driesch, D.;
        W&#246;tzel, D., and Pfaff 2007</li>
      <li>NetGenerator Non-Linear Model: &#8220;The NetGenerator Algorithm: 
        Reconstruction of Gene Regulatory Networks&#8221; of T&#246;pfer, S.; Guthke,
        R.; Driesch, D.; W&#246;tzel, D., and Pfaff 2007</li>
      <li>S-Systems: Tournier, L.: &#8220;Approximation of dynamical systems using 
        S-systems theory: application to biological systems&#8221;</li>
      <li>Hill-Radde equation: &#8220;Modeling Non-Linear Dynamic Phenomena in 
        Biochemical Networks&#8220; of Nicole Radde and Lars Kaderali</li>
      <li>Vohradsky: &#8220;Neural network model of gene expression.&#8221; of
        Vohradsk&#253;, J. 2001</li>
      <li>Weaver: &#8220;Modeling regulatory networks with weight matrices&#8221; of
        Weaver, D.; Workman, C., and Stormo, G. 1999</li>
      </ul></p>
    <p>The following new generalized kinetic equations for metabolic reactions were also
      implemented in SBMLsqueezer, which are all defined by Liebermeister et al., 
      submitted manuscript 2010:<ul>
        <li>Common modular rate law (CM)</li>
        <li>Direct binding modular rate law (DM)</li>
        <li>Force-dependent modular rate law (FM)</li>
        <li>Power-law modular rate law (PM)</li>
        <li>Simultaneous binding modular rate law (SM).</li>
      </ul>
      Each one of these equations is only available in a reversible mode and the
      user must choose one of the three possible versions that determine the degree of
      thermodynamic correctnes. The <span style="font-family:Courier">weg</span> 
      version is the most complicated form but always thermodynamically correct. 
      The <span style="font-family:Courier">hal</span> version ensures thermodynamic
      correctness in most cases, whereas for the 
      <span style="font-family:Courier">cat</span> version thermodynamic properties
      cannot be guaranteed to be correct. The versions can be selected in the preferences
      dialog or provides as a command-line option.</p>
    <p>Finally, the generalized form of Hill's equation as proposed by Cornish-Bowden
      (2004, p. 314 in &#8220;Fundamentals of Enzyme Kinetics&#8221;) has been implemented
      and is now also available for metabolic reactions.</p>
    <p>If the corresponding flag is set, boundary conditions are also set for species that
      represent empty sets, i.e., degraded species, sources, or sinks in a reaction.</p>
      
    <h3>Support and Assignment of Units and Unit Definitions</h3>
    
    <p>For multy-compartment models concentration versus molecule counts matters. 
      SBMLsqueezer now provides two different ways of ensuring unit correctness. First,
      each species can be brought to units of concentration. To this end, SBMLsqueezer
      interpretes the <span style="font-family:Courier">hasOnlySubstanceUnits</span>
      attribute of each species. Depending on whether this is set to true, the species
      is divided by the size of the surrounding compartment. Otherwise the species already
      represents a concentration and nothing is done. Second, all species can be brought
      to molecule count units. This is again achieved with the help of the <span 
      style="font-family:Courier">hasOnlySubstanceUnits</span> attribute, but now
      SBMLsqueezer multiplies with the compartment size if this attribute is false.</p>
    <p>In contrast to earlier versions, SBMLsqueezer now equipps all newly created
      <span style="font-family:Courier">Parameter</span> objects with units. To this
      end, SBMLsqueezer derives the correct units from the context. For instance, in
      the mass action kinetics the units of the catalytic constants depend on the order
      of the reaction and have therefore to be computed whenever such a rate law is
      created. Furthermore, SBMLsqueezer also to considers the <span 
      style="font-family:Courier">hasOnlySubstanceUnits</span> property of each species
      and the way in which the surrounding compartment comes into play to derive the
      units of parameters.</p>
    <p>To make models more realistic, SBMLsqueezer redefines the pre-defined SBML unit
      definition <span style="font-family:Courier">substance</span> from mol to mmol and
      <span style="font-family:Courier">volume</span> from l to ml.</p>
    <p>If possible, SBMLsqueezer avoids creating new <span
      style="font-family:Courier">UnitDefinition</span> objects. It first tries to find
      equivalent and already existing definitions in the model.</p>
      
    <h2>Bug Fixes</h2>
    
    <p>In the original paper of Hinze et al. 2007, inhibition was expressed with
      1 - product of all inhibition functions. Earlier versions of SBMLsqueezer
      created a product that run over all inhibitors with (1 - inhibition function)
      for each such function.</p>
    <p>In all gene-regulation kinetics, it is now assumed that degradation processes
      are distinct reactions that can be, e.g., modeled using a mass action rate law.
      Hence, no degradation terms can be found in gene-regulation kinetics anymore.</p>
    <p>The rate law corresponging to SBO:0000266 was not correct. Earlier versions of
      SBMLsqueezer produced the fraction Kmjs/Kibj * I, but correct is S * I/Kibj (see
      Equation 10 in the <a 
      href="http://www.biomedcentral.com/content/supplementary/1752-0509-2-39-s1.pdf">
      supplementary material &#8220;Kinetic Laws&#8221;</a>; the correct version of
      this equation can be found at <a href="http://www.ebi.ac.uk/sbo/main/SBO:0000266">
      http://www.ebi.ac.uk/sbo/main/SBO:0000266</a>).</p>
    <p>In case of an explicit enzyme catalysis, erlier versions of SBMLsqueezer did not
      multiply the numerator of equation <a 
      href="http://www.ebi.ac.uk/sbo/main/SBO:0000273">SBO:0000273</a> with the enzyme.
      Therfore, it always produced the Vmax version of the equation, but the parameter
      was called kcat.</p>
      
    <h2>References</h2>
    
    <p>Parts of the content in these release notes have been taken from the PhD 
      thesis of Andreas Dr&#228;ger (2010), University of T&#252;bingen, Germany.</p>
      
  </body>
</html>